kie-assets-plugin/src/test/java/org/kie/mojos/config/ActiveMojoSetupTest.java

```diff
@@ -0,0 +1,313 @@
+                .setProjectStructures(Arrays.asList(getStructure("s1"), getStructure("s2"), getStructure("s3")))
+                .setActiveDefinitions(Collections.singleton("!d2"));
+        List<String> combinations = new ArrayList<>();
+        activeMojoSetup.apply((definition, structure) -> {
+            combinations.add(definition.getId() + "|" + structure.getId());
+        });
+        Assert.assertThat(combinations.size(), Matchers.equalTo(6));
+        Assert.assertThat(combinations, Matchers.contains(
+                "d1|s1", "d1|s2", "d1|s3",
+                "d3|s1", "d3|s2", "d3|s3"));
+    }
+
+    @Test
+    public void testNegationSingleActiveStructure() {
+        ActiveMojoSetup activeMojoSetup = new ActiveMojoSetup()
+                .setProjectDefinitions(Arrays.asList(getDefinition("d1"), getDefinition("d2"), getDefinition("d3")))
+                .setProjectStructures(Arrays.asList(getStructure("s1"), getStructure("s2"), getStructure("s3")))
+                .setActiveStructures(Collections.singleton("!s3"));
+        List<String> combinations = new ArrayList<>();
+        activeMojoSetup.apply((definition, structure) -> {
+            combinations.add(definition.getId() + "|" + structure.getId());
+        });
+        Assert.assertThat(combinations.size(), Matchers.equalTo(6));
+        Assert.assertThat(combinations, Matchers.contains(
+                "d1|s1", "d1|s2",
+                "d2|s1", "d2|s2",
+                "d3|s1", "d3|s2"));
+    }
+
+    @Test
+    public void testNegationCombination() {
+        ActiveMojoSetup activeMojoSetup = new ActiveMojoSetup()
+                .setProjectDefinitions(Arrays.asList(getDefinition("d1"), getDefinition("d2"), getDefinition("d3")))
+                .setProjectStructures(Arrays.asList(getStructure("s1"), getStructure("s2"), getStructure("s3")))
+                .setActiveDefinitions(Collections.singleton("!d1"))
+                .setActiveStructures(Collections.singleton("!s3"));
+        List<String> combinations = new ArrayList<>();
+        activeMojoSetup.apply((definition, structure) -> {
+            combinations.add(definition.getId() + "|" + structure.getId());
+        });
+        Assert.assertThat(combinations.size(), Matchers.equalTo(4));
+        Assert.assertThat(combinations, Matchers.contains(
+                "d2|s1", "d2|s2",
+                "d3|s1", "d3|s2"));
+    }
+
+    @Test
+    public void testNegativeAndPositiveCombination() {
+        ActiveMojoSetup activeMojoSetup = new ActiveMojoSetup()
+                .setProjectDefinitions(Arrays.asList(getDefinition("d1"), getDefinition("d2"), getDefinition("d3")))
+                .setProjectStructures(Arrays.asList(getStructure("s1"), getStructure("s2"), getStructure("s3")))
+                .setActiveDefinitions(new HashSet<>(Arrays.asList("d1", "!d2")))
+                .setActiveStructures(new HashSet<>(Arrays.asList("!s1", "s3")));
```

Comment:
Yes,correct. Once there's a negation as part of the activating properties, the mode is following:
1. iterate through all the ids
2. if the id is in the activation property as a negation, skip it.
This accounts for the fact that once there's any negation the expression will work as a complement ( meaning from `a,b,c` if I specify `!b`, it means `a,c` -> all those that are not negated). So the positive selections are irrelevant.

When there are just positive activation expressions, the mode is obvious.

I did not do a formal proof, but I did not see any case where this would be breaking things (once you agree that `d1,!d1` is effectively `!d1`).

Reply:
I am adding test cases for multiple negation expressions.

Reply:
Ok, there's one not so obvious catch - if we specify negation of a non-existent id, then everything gets included. Adding a test case to illustrate `testNegationNonExistingActiveDefinitions`